### DISCUSSION TITLE
Metal build with Swift macOS app?

### DISCUSSION THREAD

**Discussion by psugihara** [ggml-org/llama.cpp#2493] [Q&A] [answered]
I have a macOS app going in Xcode that starts the `server` binary as a subprocess and hits it on localhost for completions. That’s working pretty well but if I try to use a binary built with `LLAMA_METAL=1` I see errors like this in the server output:

```
ggml_metal_init: using MPS ggml_metal_init: loading '(null)'
ggml_metal_init: error: Error Domain=NSCocoaErrorDomain Code=258 "The file name is invalid."
```

It looks like it’s not finding the metal file. I’ve tried including ggml-metal.metal both as a source file and as a resource but I still just see the above errors.

Has anyone had luck with this or know what I’m missing? Thanks!!

**Comment by guinmoon** [comment, marked as answer]
Try add `ggml-metal.metal` to metal dir, add this dir to resourses and in `ggml-metal.m` change path to this:
```
NSString *path = [NSBundle.mainBundle.resourcePath stringByAppendingString:@"/metal/ggml-metal.metal"];
```
This worked in my [swiftui app](https://github.com/guinmoon/LLMFarm).
Or use absolute path like this:
```
NSString *path = @"/Users/guinmoon/dev/alpaca_llama_etc/LLMFarm/metal/ggml-metal.metal";
```
